lib/model/sharedpullerstate.go

```diff
@@ -265,18 +265,20 @@ func (s *sharedPullerState) finalClose() (bool, error) {
 		return false, nil
 	}
 
-	if s.fd != nil {
-		if err := s.fd.Sync(); err != nil {
+	if s.writer != nil {
+		s.writer.mut.Lock()
+		if err := s.writer.fd.Sync(); err != nil {
 			// Sync() is nice if it works but not worth failing the
 			// operation over if it fails.
 			l.Debugf("fsync %q failed: %v", s.tempName, err)
 		}
 
-		if err := s.fd.Close(); err != nil && s.err == nil {
+		if err := s.writer.fd.Close(); err != nil && s.err == nil {
 			// This is our error as we weren't errored before.
 			s.err = err
 		}
-		s.fd = nil
+		s.writer.mut.Unlock()
+		s.writer = nil
 	}
```

Comment:
Why are we changing shit here? The only reason locking exists on the write is to prevent interleaved writes, otherwise no locking is needed?